keps/sig-node/0035-20190130-topology-manager.md

```diff
@@ -313,6 +313,101 @@ _Figure: Topology Manager hint provider registration._
 
 _Figure: Topology Manager fetches affinity from hint providers._
 
+Additionally, we propose an extension to the device plugin interface to help
+influence `TopologyHint` generation, as well as influence overall allocation
+decisions made by the `devicemanager`. The diff below shows the proposed
+changes:
+
+```
+diff --git a/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto b/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto
+index 758da317fe..1e55d9c541 100644
+--- a/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto
++++ b/pkg/kubelet/apis/deviceplugin/v1beta1/api.proto
+@@ -55,6 +55,11 @@ service DevicePlugin {
+    // returns the new list
+    rpc ListAndWatch(Empty) returns (stream ListAndWatchResponse) {}
+
++   // GetPreferredAllocations returns a List of preferred Device Allocations
++   // given the current state of allocated devices and an allocation request
++   // for a specific number of new devices.
++   rpc GetPreferredAllocations(PreferredAllocationsRequest) returns (PreferredAllocationsResponse) {}
++
+    // Allocate is called during container creation so that the Device
+    // Plugin can run device specific operations and instruct Kubelet
+    // of the steps to make the Device available in the container
+@@ -99,6 +104,31 @@ message PreStartContainerRequest {
+ message PreStartContainerResponse {
+ }
+
++// - PreferredAllocationsRequest is expected to be passed via a call to
++//   `GetPreferredAllocations()` at pod admission time.
++// - Its results will be used to help generate TopologyHints for the
++//   TopologyManager, as well as make more informed allocation decisions when
++//   device allocation is performed by the `devicemanager`.
++// - It is assumed that the device plugin is not aware of the current
++//   allocation state, so all available devices should be passed, via the
++//   `available_deviceIDs` field.
++// - Likewise, any devices that must be included as part of each allocation
++//   (e.g. because of init-container inheritance) should be passed via the
++//   `must_include_deviceIDs` fields.
++// - The `size` field indicates the size (i.e. total number of devices) that
++//   should be included in each preferred allocation response. This size
++//   includes any devices that are part of the `must_include_deviceIDs` field.
++message PreferredAllocationsRequest {
++   repeated string available_deviceIDs = 1;
++   repeated string must_include_deviceIDs = 2;
++   int32 size = 3;
++}
```

Comment:
Does it make sense to pass the `TopologyHint` here as well, so that the plugin can condense the set of `PreferredAllocations` it sends back -- or is this better left to the `decicemanager` to merge this information?

Reply:
Outdated, based on other feedback.